third_party/move/move-vm/runtime/src/data_cache.rs

```diff
@@ -19,17 +19,17 @@ use move_vm_types::{
     loaded_data::runtime_types::Type,
     values::{GlobalValue, Value},
 };
-use std::collections::btree_map::BTreeMap;
+use std::collections::{btree_map::BTreeMap, hash_map::HashMap};
 
 pub struct AccountDataCache {
-    data_map: BTreeMap<Type, (MoveTypeLayout, GlobalValue)>,
+    data_map: HashMap<Type, (MoveTypeLayout, GlobalValue)>,
```

Comment:
There's a funny reason behind this: BTreeMap requires key to have `Ord` whereas HashMap requires `Hash`. After we add ability information to `Type`, those ability information need to be erased when querying from the map here.

Reply:
Based on some of the experiments we did a long time ago, I'd actually not assume `HashMap` would be faster than `BTreeMap`. The reason is that for regular inputs, `BTreeMap` almost never has to perform full comparisons while `HashMap` requires the key to be fully hashed.

Reply:
For tree-structured data, we may want to do something to avoid recursive Hash computation.  I was just investigating using something like https://docs.rs/by_address/latest/by_address/# for SignatureToken recursive links to avoid having to recompute hashes for subtrees.

But my experiments showed that for sizes < 1000 BTreeMap was better than HashMap for random integer data.  Above that HashMap starts to win but still <2x.

Reply:
@wrwg That's useful to know. In my experience (C++ tho). There was never a good way to predict whether to use Tree or Hash based map - it just depends on so many things.